File: apps/share-backend/src/store.ts
/**
 * Persistence interface for the rendezvous backend. The router talks
 * to this; the production binding is KV-backed, tests use an
 * in-memory fake. Splitting the contract here keeps handlers unaware
 * of Cloudflare-isms and makes the storage layer swappable (KV today,
 * Durable Objects tomorrow if we need strong single-pickup).
 */

export type RendezvousRecord = {
	/** Recipient's ephemeral X25519 public key. */
	readonly epk: Uint8Array;
	/** Unix seconds when this rendezvous (and its blob) expire. */
	readonly expiresAt: number;
};

export interface RendezvousStore {
	/**
	 * Insert a fresh rendezvous. Resolves to `true` if the code was
	 * unused; `false` if a rendezvous already exists under it (the
	 * caller should reject the request — a code collision inside the
	 * TTL window is a 60-bit pre-image hit and means re-derive).
	 */
	putRendezvous(
		code: string,
		record: RendezvousRecord,
		ttlSeconds: number,
	): Promise<boolean>;

	/**
	 * Look up the recipient's epk for `code`. Returns `null` only if
	 * the record is truly absent; an expired-but-still-stored record
	 * is returned and the handler decides between `404` and `410`.
	 * Collapsing both states inside the store would lose the
	 * documented "missing vs. expired" distinction in the API.
	 */
	getRendezvous(code: string): Promise<RendezvousRecord | null>;

	/**
	 * Stage the encrypted blob under `code`. Resolves `false` if a
	 * blob is already present — uploads are single-shot (ADR 0007).
	 */
	putBlob(code: string, blob: Uint8Array, ttlSeconds: number): Promise<boolean>;

	/**
	 * Atomically read-and-delete the blob for `code`. Returns `null`
	 * if the blob is missing — including the case where another
	 * recipient just picked it up.
	 */
	takeBlob(code: string): Promise<Uint8Array | null>;

	/**
	 * Increment the per-IP mint counter inside the current TTL window.
	 * Returns the post-increment count so the caller can decide
	 * whether to throttle.
	 */
	incrementMintCounter(ip: string, ttlSeconds: number): Promise<number>;
}

File: apps/share-backend/src/store.rs
//! Persistence trait for the rendezvous backend.
//!
//! Splits transport from storage so the in-memory impl
//! (single-replica Knative) and any future shared-store impl (Redis,
//! Postgres) plug in behind the same surface.

use std::sync::Arc;

/// The recipient's ephemeral X25519 public key + expiry timestamp.
#[derive(Debug, Clone)]
pub struct RendezvousRecord {
    pub epk: Vec<u8>,
    pub expires_at: u64,
}

/// `Send + Sync` is required because the store is held inside an
/// `Arc` shared across axum's worker tasks.
pub trait RendezvousStore: Send + Sync + std::fmt::Debug {
    /// Insert a fresh rendezvous. Returns `true` if the code was
    /// unused. `false` means the same 12-char code is already taken
    /// inside the TTL window — a real 60-bit pre-image collision;
    /// the caller should ask the client to retry with a fresh epk.
    /// Sweep of stale entries runs on insert so the map stays small.
    fn put_rendezvous(&self, code: &str, record: RendezvousRecord) -> bool;

    /// Look up a record by code. **Returns the record even if
    /// expired** so the handler can distinguish `410 Gone` from
    /// `404 Not Found` — eviction happens on writes only, so a
    /// just-expired record is still visible here until the next
    /// mutation.
    fn get_rendezvous(&self, code: &str) -> Option<RendezvousRecord>;

    /// Stage the encrypted blob under `code` with an absolute
    /// `expires_at` (unix seconds). Returns `false` if a blob is
    /// already present — uploads are single-shot per ADR 0007. The
    /// in-memory impl gives us a real atomic insert-if-absent under
    /// a `Mutex`, which the previous KV+R2 pair couldn't.
    fn put_blob(&self, code: &str, blob: Vec<u8>, expires_at: u64) -> bool;

    /// Atomically read-and-delete the blob for `code`. The caller
    /// passes `now` (unix seconds, from the handler's injectable
    /// clock) so an expired blob is rejected here rather than
    /// relying on the background sweep — the sweep is best-effort.
    /// Returns `None` if missing, already-picked-up, or expired.
    fn take_blob(&self, code: &str, now: u64) -> Option<Vec<u8>>;

    /// Increment the per-IP mint counter inside the current TTL
    /// window. Returns the post-increment count so the handler can
    /// throttle.
    fn increment_mint_counter(&self, ip: &str, ttl_seconds: u64) -> u32;
}

pub type SharedStore = Arc<dyn RendezvousStore>;
